jetty-io/src/main/java/org/eclipse/jetty/io/RetainableByteBuffer.java

```diff
@@ -15,62 +15,71 @@
 
 import java.nio.ByteBuffer;
 import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.function.Consumer;
 
 import org.eclipse.jetty.util.BufferUtil;
 import org.eclipse.jetty.util.Retainable;
 
 /**
  * A Retainable ByteBuffer.
- * <p>Acquires a ByteBuffer from a {@link ByteBufferPool} and maintains a reference count that is
+ * <p>A ByteBuffer which maintains a reference count that is
  * initially 1, incremented with {@link #retain()} and decremented with {@link #release()}. The buffer
  * is released to the pool when the reference count is decremented to 0.</p>
  */
 public class RetainableByteBuffer implements Retainable
 {
-    private final ByteBufferPool pool;
     private final ByteBuffer buffer;
     private final AtomicInteger references;
+    private final Consumer<ByteBuffer> releaser;
+    private final AtomicLong lastUpdate = new AtomicLong(System.nanoTime());
 
-    public RetainableByteBuffer(ByteBufferPool pool, int size)
+    RetainableByteBuffer(ByteBuffer buffer, Consumer<ByteBuffer> releaser)
     {
-        this(pool, size, false);
+        this.releaser = releaser;
+        this.buffer = buffer;
+        this.references = new AtomicInteger(1);
     }
 
-    public RetainableByteBuffer(ByteBufferPool pool, int size, boolean direct)
+    public int capacity()
     {
-        this.pool = pool;
-        this.buffer = pool.acquire(size, direct);
-        this.references = new AtomicInteger(1);
+        return buffer.capacity();
     }
 
     public ByteBuffer getBuffer()
     {
         return buffer;
     }
 
+    public long getLastUpdate()
+    {
+        return lastUpdate.getOpaque();
+    }
+
     public int getReferences()
     {
         return references.get();
     }
 
+    public boolean isDirect()
+    {
+        return buffer.isDirect();
+    }
+
     @Override
     public void retain()
     {
-        while (true)
-        {
-            int r = references.get();
-            if (r == 0)
-                throw new IllegalStateException("released " + this);
-            if (references.compareAndSet(r, r + 1))
-                break;
-        }
+        references.incrementAndGet();
```

Comment:
I would retain the safety code.

Reply:
Okay, I could add it back after a little logic change that requires discriminating references increments due to user-code or due to pool code.